### ISSUE TITLE
Add support for dynamic test instances

### ISSUE THREAD

**Issue by mfulton26** [junit-team/junit5#638]
Dynamic tests can be generated by defining test factories using `@TestFactory` but these methods cannot be used to define groups of tests or to dynamically generate nested tests. Adding support for dynamic test instances would allow defining test instance factories containing tests, test factories, and/or nested test instance factories.

e.g.

```java
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;
import static org.junit.jupiter.api.DynamicTest.dynamicTest;
import static org.junit.jupiter.api.DynamicTestInstance.dynamicTestInstance; // proposed type

import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.TestFactory;
import org.junit.jupiter.api.TestInstanceFactory; // proposed annotation
import org.junit.jupiter.api.function.Instantiatable; // proposed interface

class DynamicTestInstancesDemo {
    // This will dynamically generate instances of `DynamicTestInstancesDemo.LocalizationTests`.
    // Each instance will have a unique `locale` associated with it.
    @TestInstanceFactory
    Stream<DynamicTestInstance> dynamicTestInstancesFromStream() {
        return Arrays.stream(Locale.getAvailableLocales()).map(locale -> {
            String displayName = locale.getDisplayName();
            Instantiatable instantiatable = () -> new LocalizationTests(locale);
            return dynamicTestInstance(displayName, instantiatable);
        });
    }

    class LocalizationTests {
        // Some data unique to each dynamic test instance of `LocalizationTests`
        final Locale locale;

        LocalizationTests(Locale locale) {
            this.locale = locale;
        }

        // This is a test within each dynamic test instance
        @Test
        void something() { ... }

        // This is another test within each dynamic test instance
        @Test
        void somethingElse() { ... }

        // This will generate tests within each dynamic test instance
        @TestFactory
        Stream<DynamicTest> dynamicTests() { ... }

        // This will generate nested test instances within each dynamic test instance
        @TestInstanceFactory
        Stream<DynamicTestInstance> dynamicTestInstances() { ... }
    }
}
```

**Comment by marcphilipp** [comment]
Interesting idea! IMHO this looks a lot like parameterized tests, doesn't it? How about we add support for those (#14) first and then see if this proposal provides additional value?

**Comment by mfulton26** [comment]
I wonder if it might make more sense to actually implement support for dynamic test instances first and then see if paramterized tests provide any value. From what I can tell (please correct me if I am wrong), paramterized tests are specialized dynamic tests and won't support grouping or nesting tests, etc.

**Comment by mfulton26** [comment]
I see that support for `@ParamterizedTest` has been merged into master. Taking a quick look at it it does seem to simplify creating tests dynamically which I find very helpful.

I have used Spek some as well as other test runners but I would very much like to use JUnit Jupiter. The main issue I am having however is that I still find places where I could really use a way to define a group of nested tests dynamically. Maybe something like a `@ParameterizedTests` as a class constructor annotation similar to `@ParamterizedTest` but for dynamically creating instances of test classes given various sources of data, etc.

**Comment by marcphilipp** [comment]
Please have a look at https://github.com/junit-team/junit5/issues/695#issuecomment-292914486. I wonder if we could address both requirements at once.

**Comment by nicolaiparlog** [comment]
@mfulton26 Isn't this very similar to #604, which is currently being solved in #813? If you agree, this issue can be closed. If you think about how to best and most flexibly create dynamic tests, chime in in #371.

**Comment by sormuras** [comment]
Duplicate of #604
